### PR TITLE
Static HTML profile page

### PR DESCRIPTION
**Pull request by valer-whylabs**
Create a static HTML page where users can load their generated profile (JSON) file.
The page should give them an insight, how the observatory profile page works - with limited functionalities.

### CODE REVIEW

File: viewer/index.html [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+
+  <body>
+    <nav class="navbar navbar-inverse fixed-top">
+      <div class="container-fluid">
+        <div class="navbar-header">
+          <h5>Profile viewer for whylogs</h5>
+          <ul class="navbar-nav d-inline-block">
+            <li class="nav-item">
+              <p>
+                Maintianed by
```
Comment (phantomgoose) [inline viewer/index.html]: typo: Maintained :)

File: viewer/index.html [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+                <a class="navbar-link" href="https://try.whylabsapp.com/" target="_blank" rel="noreferrer noopener"
```
Comment (phantomgoose) [inline viewer/index.html]: we may want to add a `?utm_source` here so we know who's coming into the sandbox from this page

File: viewer/index.html [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+            </div>
+          </div>
+        </div>
+      </div>
+    </div>
+
+    <!-- Bootstrap core JavaScript
+    ================================================== -->
+    <!-- Placed at the end of the document so the pages load faster -->
+    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js"></script>
```
Comment (phantomgoose) [inline viewer/index.html]: do we need jquery for this?

File: viewer/index.html (hunk 2 above)
Comment (mrmcduff) [reply viewer/index.html]: what would the utm_source be? Should we make a special tag for that?

File: viewer/index.html (hunk 3 above)
Comment (mrmcduff) [reply viewer/index.html]: It's used in the JS -- sort of our replacement interactivity (since there is no react)

File: viewer/index.html [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+    <script src="https://d3js.org/d3.v6.min.js"></script>
+    <script
```
Comment (mrmcduff) [inline viewer/index.html]: @valer-whylabs  -- Just realized; we can't call out to external libraries like this, so we should download the minified versions of them from
https://jquery.com/download/
https://d3js.org/
and
https://getbootstrap.com/docs/4.3/getting-started/download/

Yet another reason why it's good that we kept the dependencies to a minimum

File: viewer/index.html (hunk 1 above)
Comment (mrmcduff) [inline viewer/index.html]:
```suggestion
                Maintained by
```

File: viewer/whylogs-script.js [hunk 5]
```diff
@@ -0,0 +1,547 @@
...
+
+(function () {
+  var MESSAGES = {
+    error: {
+      noInputElementFound: "It seems we could not find the file input element.",
+      noBrowserSupport: "This browser does not seem to support the `files` property of file inputs.",
+      noFileSelected: "Please select a file.",
+      fileAPINotSupported: "The file API is not supported on this browser yet.",
+      invalidJSONFile:
+        "It seems the JSON file you are trying to load is not valid due because it does not match whylogs format. Please check the file validity and try again.",
```
Comment (mrmcduff) [inline viewer/whylogs-script.js]:
```suggestion
        "The JSON file you are trying to load is does not match the expected whylogs format. Please check the file and try again.",
```

File: viewer/whylogs-script.js [hunk 6]
```diff
@@ -0,0 +1,547 @@
...
+  }
+
+  function scrollToFeatureName(event) {
+    var TABLE_HEADER_OFFSET = 90;
+    var $tableWrap = $(".wl-table-wrap");
+    var featureNameId = event.currentTarget.dataset.featureNameId;
+    var currentOffsetTop = $tableWrap.scrollTop();
+    var offsetTop = $('[data-feature-name="' + featureNameId + '"]').offset().top;
+
+    // $tableWrap.scrollTop(currentOffsetTop + offsetTop - TABLE_HEADER_OFFSET);
```
Comment (mrmcduff) [inline viewer/whylogs-script.js]: Remove commented line!

**Review by mrmcduff — CHANGES_REQUESTED** [review]
Great work! We do need to address a few of the suggestions (typo / text changes) and download the minified versions of the external libraries.

File: viewer/index.html (hunk 2 above)
Comment (phantomgoose) [reply viewer/index.html]: i think it doesn't matter, as long as it doesn't change (very often). Sam might have better ideas, but we could do something like `whylogs_static_html_page` even. Im honestly not sure what the best practices around this are, I just know we track this in Heap :)

File: viewer/whylogs-script.js [hunk 7]
```diff
@@ -0,0 +1,547 @@
...
+      noBrowserSupport: "This browser does not seem to support the `files` property of file inputs.",
+      noFileSelected: "Please select a file.",
+      fileAPINotSupported: "The file API is not supported on this browser yet.",
+      invalidJSONFile:
+        "It seems the JSON file you are trying to load is not valid due because it does not match whylogs format. Please check the file validity and try again.",
+    },
+  };
+
+  // HTML Elements
+  var $selectedProfile = $(".wl__selected-profile");
```
Comment (phantomgoose) [inline viewer/whylogs-script.js]: is there a reason to use `var` keyword in this file?

File: viewer/whylogs-script.js [hunk 8]
```diff
@@ -0,0 +1,547 @@
...
+  // Constants and variables
+  var batchArray = [];
+  var featureSearchValue = "";
+  var isActiveInferredType = {};
+  var propertyPanelData = [];
+  var profiles = [];
+  var jsonData = {};
+
+  // Util functions
+  function debounce(func, wait, immediate) {
```
Comment (phantomgoose) [inline viewer/whylogs-script.js]: cant we use lodash? or trying to avoid dependencies?

File: viewer/whylogs-script.js (hunk 5 above)
Comment (phantomgoose) [reply viewer/whylogs-script.js]: There's an extra `is` here :)

`The JSON file you are trying to load does not match the expected whylogs format. Please check the file and try again`

File: viewer/index.html (hunk 3 above)
Comment (phantomgoose) [reply viewer/index.html]: You guys probably saw this already, but it looks like there's a library for generating static pages with react: https://github.com/react-static/react-static

But yea, jquery is probably a bit easier to work with than vanilla js in this case (especially if we're trying to keep dependencies to a minimum).

File: viewer/index.html (hunk 2 above)
Comment (samgracie) [reply viewer/index.html]: I was thinking something like that, Alex. Let's set the tag to `whylogs_html_viewer`, it will be an important metric to break out.

File: viewer/index.html (hunk 4 above)
Comment (samgracie) [reply viewer/index.html]: Should we include the full libraries or extract the specific parts we need for the html viewer?

File: viewer/index.html (hunk 3 above)
Comment (valer-whylabs) [reply viewer/index.html]: We were discussing react-static, but because we already had something in vanilla and in order to keep dependencies at the minimum, we decided to continue without vanilla.

File: viewer/whylogs-script.js (hunk 7 above)
Comment (valer-whylabs) [reply viewer/whylogs-script.js]: There is no specific reason for that. I started writing code with ES5 because we are not using any compiler/transpiler and I wanted to have a greater browser support. However, every required browser now supports ES6 by default, so it was my silly mistake.

I would probably recommend a quick refactoring to ES6, because we are mixing arrow functions and regular functions as well. Just to be more consistent.

File: viewer/whylogs-script.js (hunk 8 above)
Comment (valer-whylabs) [reply viewer/whylogs-script.js]: I can import only `debounce` from Lodash as it will be 2-3 functions only, nothing heavy and complex. However, I was trying to avoid external libraries if they are not necessary.

Comment (phantomgoose) [reply viewer/whylogs-script.js]: i think we'll end up having to write a lot of helper functions, especially if we're just sticking with vanilla/jquery. Adding lodash prolly makes sense, and gives us peace of mind that these utils are implemented correctly and dont need tests, so we can just focus on the page itself.

File: viewer/whylogs-script.js (hunk 7 above)
Comment (phantomgoose) [reply viewer/whylogs-script.js]: +1 for refactoring to ES6 for readability and consistency!

File: viewer/index.html (hunk 4 above)
Comment (mrmcduff) [reply viewer/index.html]: It's all local; I'm not sure what the size is nor how difficult it is to extract

Comment (valer-whylabs) [reply viewer/index.html]: The whole `viewer` directory takes up 4.5MB on the disk. We can reduce it by removing some of the font types, as they only take up 2.2MB.

File: viewer/whylogs-script.js (hunk 8 above)
Comment (valer-whylabs) [reply viewer/whylogs-script.js]: I would leave it as is for now, because we are using only the debounce util function. Once we need one more function I will import Lodash. What do you think?

Comment (mrmcduff) [reply viewer/whylogs-script.js]: I'm inclined to be fine for this release with just the one instance -- it's important to keep in mind that we will update this as more functionality becomes available.

**Review by mrmcduff — APPROVED** [review]

**Review by lalmei — APPROVED** [review]
LGTM
